feat: upload_directory_cnt() for ConnectorFS

akselthomsen opened the pull request:
## Summary
Added `upload_directory_cnt()` method for `ConnectorFS` so the behaviour of the parent FS connector is consistent with the Sharepoint connector.

In order to be able to rename the directory when copying I had to import the `{fs}` package.
If a solution without it can be found it might be better, but all I could think of involved copying the directory twice which I want to avoid.

Since `{fs}` is now imported I changed the other relevant `ConnectorFS` methods to use functions from here, since they have better error handling (fails consistently) than the base functions for manipulating files.

## Changes Made
- fix: remove dead references to connector.logger
- feat: new upload_directory_cnt method for file storage
- feat: use fs functions for other connectorFS methods

## Testing
- Added new unit test in `test-fs.R`

## Checklist
- [X] Code changes have been tested
- [X] Documentation has been updated, if applicable
- [X] All automated tests pass
- [X] Coding style and naming conventions have been followed
- [X] The PR is ready for review and merge

vladimirobucina commented on R/fs_methods.R:
```diff
@@ -263,7 +268,22 @@ remove_directory_cnt.ConnectorFS <- function(connector_object, name, ...) {
+#'
+#' @rdname upload_directory_cnt
+#' @export
+upload_directory_cnt.ConnectorFS <- function(connector_object, dir, name, ...) {
```
can this function also have `open` parameter, to get the connector with path defined as uploaded directory @falgreen

vladimirobucina replied:
Also should we have overwrite parameter?

akselthomsen replied:
yeah, we could add the extra parameters as well to the functions.

@Cervangirard and I just wanted to have the `...`and then make the user read our documentation where it says which underlying function is used, but maybe we should include parameters...
If we do, I just think we should include all parameters instead of cherry-picking from function to function

@falgreen what do you say?

akselthomsen replied:
> can this function also have `open` parameter, to get the connector with path defined as uploaded directory @falgreen

I could add it, but I think we should then have the default `open = FALSE`, since I imagine the users will most commonly not need to connect the directory they just uploaded.

vladimirobucina commented on R/fs_methods.R:
```diff
@@ -225,11 +227,14 @@ create_directory_cnt.ConnectorFS <- function(
+  # create a new connector object from the new path with persistent extra class
   if (open) {
-    connector_object <- ConnectorFS$new(path)
+    extra_class <- class(connector_object)
+    extra_class <- utils::head(extra_class, which(extra_class == "ConnectorFS") - 1)
+    connector_object <- connector_fs(path, extra_class)
```
Good catch 👍

vladimirobucina reviewed (changes requested):
It looks good, but I just have a couple of questions and suggestions :)

akselthomsen commented on R/cnt_logger_add_logs.R:
```diff
@@ -14,8 +14,7 @@
 #' 1. Checks if the input `connections` is of class "connectors".
-#' 2. Registers the S3 methods for logging using `connector.logger::cnt_logger_S3_register()`.
-#' 3. Iterates through each connection in the list and prepends the "ConnectorLogger" class.
+#' 1. Iterates through each connection in the list and prepends the "ConnectorLogger" class.
```
It is easier to always use 1. since they just get translated anyway in the Rd:
``` latex
\enumerate{
\item Checks if the input \code{connections} is of class "connectors".
\item Iterates through each connection in the list and prepends the "ConnectorLogger" class.
}
```

akselthomsen commented:
@vladimirobucina ready for review again :)
I have updated with:
- `overwrite` parameter in relevant functions that are controlled by a central option
- added `open` parameter to `upload_directory()`
- added `download_directory_cnt()` function
- updated the documentation and the methods vignette with these two new functions and fixed a few places where it was not in sync with the code

vladimirobucina commented on R/fs_write.R:
```diff
@@ -27,16 +28,29 @@ write_file <- function(x, file, ...) {
 #' @export
-write_ext.txt <- function(file, x, ...) {
+write_ext.txt <- function(file, x, overwrite = FALSE, ...) {
+  check_file_exists(file, overwrite, ...)
```
Just move this to `write_file` function instead of having it in every method.

vladimirobucina commented on vignettes/articles/Consistent-API.Rmd:
```diff
@@ -32,10 +32,12 @@ Below is a table summarizing the methods, their return values, and explanations
 | `create_directory_cnt()` | `Connector` object, if `open = TRUE`, then new `Connector` object | Returns the current `Connector` object or a new one if the directory was opened, allowing for immediate use of the new directory. |
 | `remove_directory_cnt()` | `Connector` object | Returns the `Connector` object to allow for method chaining and to confirm the directory removal was completed. |
+| `upload_directory_cnt()` | `Connector` object, if `open = TRUE`, then new `Connector` object | Returns the current `Connector` object or a new one if the directory was opened, allowing for immediate use of the uploaded directory. |
+| `download_directory_cnt()` | `dir` | Returns the path to the downloaded directory allowing for immediate use |
```
Don't forget to change this one as well.

akselthomsen replied:
Also done, but did not add the `to ensure download was completed` part, since that would require the file and not the connector_object.

vladimirobucina reviewed (changes requested):
It looks great now, just couple minor changes.

akselthomsen commented on R/fs_methods.R:
```diff
@@ -147,22 +161,24 @@ download_cnt.ConnectorFS <- function(
 
   return(
-    invisible(connector_object)
+    invisible(file)
```
Done

vladimirobucina reviewed (approved):
Everything looks good
